### PR TITLE
Fix pyenv shim fix

### PR DESCRIPTION
**Pull request by ls-jad-elkik**
The pyenv installation is lacking the last steps which are `echo`-ed here: https://raw.githubusercontent.com/pyenv/pyenv-installer/master/bin/pyenv-installer

Without this, the shims are not updated when installing a package. You'd have to run `pyenv rehash` manually after `pip install`-ing a package.

(Only updating the 3.9 Dockerfile here but I think all of them should be updated?

### CODE REVIEW

**Comment by felicianotech** [comment]
@dsayling Thoughts here?

@ls-jad-elkik Thank you for the PR. When updating Convenience Images, we only update the `Dockerfile.template` file at the root of the repository. Then, those changes get pulled into future releases. The Dockerfiles for existing releases are only there as a record of what was in that image release.

**Comment by ls-jad-elkik** [comment]
Right! I thought it looked redundant but somehow missed the template at the root directory.

I'll update this PR if needed based on the feedback about the actual change.

**Comment by dsayling** [comment]
@ls-jad-elkik can you provide an example of what you're seeing.

The pyenv installation uses the https://pyenv.run short URL which returns
```
#!/bin/bash
#
# Usage: curl https://pyenv.run | bash
#
# For more info, visit: https://raw.githubusercontent.com/pyenv/pyenv-installer
#
index_main() {
    set -e
    curl -s -S -L https://raw.githubusercontent.com/pyenv/pyenv-installer/master/bin/pyenv-installer | bash
}

index_main
```
So the last few lines of the script should still be executed. 

If you run the docker image locally, `docker run --rm -it cimg/python:3.9 /bin/bash`, what do you see that you can't do without these changes?

**Comment by ls-jad-elkik** [comment]
That script just echoes the commands that someone would need to add to their own profile, it doesn't actually modify the `.bashrc`/profile. It just echoes this to stderr (you can see it if you build the image with `--no-cache`:

```
# Load pyenv automatically by adding
# the following to ~/.bashrc:

export PATH="/home/circleci/.pyenv/bin:$PATH"
eval "$(pyenv init -)"
eval "$(pyenv virtualenv-init -)"
```

Here's a "simplified" version of the issue:

```
$ docker run --rm -it cimg/python:3.9 /bin/bash
circleci@8a75d3213940:~/project$ python -m pip install flask
Collecting flask
  Downloading Flask-1.1.2-py2.py3-none-any.whl (94 kB)
     |████████████████████████████████| 94 kB 2.0 MB/s
Collecting click>=5.1
  Downloading click-7.1.2-py2.py3-none-any.whl (82 kB)
     |████████████████████████████████| 82 kB 2.6 MB/s
Collecting Jinja2>=2.10.1
  Downloading Jinja2-2.11.3-py2.py3-none-any.whl (125 kB)
     |████████████████████████████████| 125 kB 7.5 MB/s
Collecting itsdangerous>=0.24
  Downloading itsdangerous-1.1.0-py2.py3-none-any.whl (16 kB)
Collecting Werkzeug>=0.15
  Downloading Werkzeug-1.0.1-py2.py3-none-any.whl (298 kB)
     |████████████████████████████████| 298 kB 13.3 MB/s
Collecting MarkupSafe>=0.23
  Downloading MarkupSafe-1.1.1-cp39-cp39-manylinux2010_x86_64.whl (32 kB)
Installing collected packages: click, MarkupSafe, Jinja2, itsdangerous, Werkzeug, flask
Successfully installed Jinja2-2.11.3 MarkupSafe-1.1.1 Werkzeug-1.0.1 click-7.1.2 flask-1.1.2 itsdangerous-1.1.0
WARNING: You are using pip version 20.2.3; however, version 21.0.1 is available.
You should consider upgrading via the '/home/circleci/.pyenv/versions/3.9.1/bin/python -m pip install --upgrade pip' command.
circleci@8a75d3213940:~/project$ flask --version
bash: flask: command not found
circleci@8a75d3213940:~/project$ pyenv rehash
circleci@8a75d3213940:~/project$ flask --version
Python 3.9.1
Flask 1.1.2
Werkzeug 1.0.1
circleci@8a75d3213940:~/project$
```

Weird thing is the same repro steps don't seem to work with my fixes either (after fixing the dumb mistakes). It looks like `pyenv` only runs `rehash` when using `pip` and not `python -m pip` - I guess that's beyond the scope of this image.

Our *actual* use case which my fix seems to fix - but maybe again I am missing something - is in our own Docker image which uses this one as a base (when we used `cimg/base` as a base image, it worked fine because pyenv was not involved).

Our Dockerfile does this basically:

```
FROM cimg/python:3.9.1

# ...

COPY install.sh /opt/install.sh
RUN sudo chmod +x /opt/install.sh
RUN /opt/install.sh
```

And `install.sh` does a bunch of stuff and eventually:

```
#!/bin/sh

set -e
set -x

# ...
python -m pip install flask
# ...
```

After doing that: `docker run --rm -it our-image which flask` would fail.

With this fix `docker run --rm -it our-image bash -ic "which flask"` succeeds. If you run it with `-x` you'll see `+++ command pyenv rehash` kicking in as part of the `.bashrc` pyenv init part. That's what actually fixes it.

Not sure what else `pyenv init` and `pyenv virtualenv-init` do - if it only updates the prompt or something like that then maybe it's not such a big deal...

**TL;DR**: it looks like just using `pip install` instead of `python -m pip install` would have fixed this issue we have - you still don't run `pyenv init` which can cause some unexpected behaviour as compared to the "vanilla" Python in `cimg/base` and the regular `pyenv` local installations.

**Comment by dsayling** [comment]
@ls-jad-elkik - I built the docker image you are proposing here are and it seems to have broken the shim links in general - I haven't really investigated beyond this, but I'm not sure this is the correct solution. I believe the 're'-initing is causing the paths to break. Meanwhile, I'll look at some best practices around `pyenv` in docker images and how `pip` works in the `pyenv` shim. 

```
testdk|⇒ docker run -it --rm cimg-python-pr76:latest cat /home/circleci/.pyenv.bashrc
export PATH="${PYENV_ROOT}/bin:$PATH"
eval "$(pyenv init -)"
eval "$(pyenv virtualenv-init -)"
```
```
testdk|⇒ cat Dockerfile 
FROM cimg-python-pr76:latest 
#FROM cimg/python:3.9.1

COPY test-install.sh /opt/install.sh
RUN sudo chmod +x /opt/install.sh
RUN /opt/install.sh

testdk|⇒ cat test-install.sh 
#!/bin/sh

set -e
set -x

#python -m pip install pytest
pip install pytest
```
```
testdk|⇒ docker build -t test1 .
Sending build context to Docker daemon  3.072kB
Step 1/4 : FROM cimg-python-pr76:latest
 ---> c87f0ea132d0
Step 2/4 : COPY test-install.sh /opt/install.sh
 ---> d2b2badb7816
Step 3/4 : RUN sudo chmod +x /opt/install.sh
 ---> Running in cd52c33cac16
+ sudo chmod +x /opt/install.sh
Removing intermediate container cd52c33cac16
 ---> 4fa9a47413db
Step 4/4 : RUN /opt/install.sh
 ---> Running in 8c49bc9af716
+ /opt/install.sh
+ pip install pytest
/opt/install.sh: 7: pip: not found
The command '/bin/bash -exo pipefail -c /opt/install.sh' returned a non-zero code: 127
```

If I use the existing image - this works without issue

```
testdk|⇒ cat Dockerfile
#FROM cimg-python-pr76:latest 
FROM cimg/python:3.9.1

COPY test-install.sh /opt/install.sh
RUN sudo chmod +x /opt/install.sh
RUN /opt/install.sh
```
```
testdk|⇒ docker build -t test2 . --no-cache
Sending build context to Docker daemon  3.072kB
Step 1/4 : FROM cimg/python:3.9.1
 ---> add8ca842db4
Step 2/4 : COPY test-install.sh /opt/install.sh
 ---> 04b25fd794cf
Step 3/4 : RUN sudo chmod +x /opt/install.sh
 ---> Running in 0f462964f9bd
+ sudo chmod +x /opt/install.sh
Removing intermediate container 0f462964f9bd
 ---> 1e32c7b7b025
Step 4/4 : RUN /opt/install.sh
 ---> Running in 13ec3c5acf52
+ /opt/install.sh
+ pip install pytest
Collecting pytest
  Downloading pytest-6.2.2-py3-none-any.whl (280 kB)
Collecting toml
  Downloading toml-0.10.2-py2.py3-none-any.whl (16 kB)
Collecting iniconfig
  Downloading iniconfig-1.1.1-py2.py3-none-any.whl (5.0 kB)
Collecting py>=1.8.2
  Downloading py-1.10.0-py2.py3-none-any.whl (97 kB)
Collecting packaging
  Downloading packaging-20.9-py2.py3-none-any.whl (40 kB)
Collecting attrs>=19.2.0
  Downloading attrs-20.3.0-py2.py3-none-any.whl (49 kB)
Collecting pluggy<1.0.0a1,>=0.12
  Downloading pluggy-0.13.1-py2.py3-none-any.whl (18 kB)
Collecting pyparsing>=2.0.2
  Downloading pyparsing-2.4.7-py2.py3-none-any.whl (67 kB)
Installing collected packages: toml, iniconfig, py, pyparsing, packaging, attrs, pluggy, pytest
Successfully installed attrs-20.3.0 iniconfig-1.1.1 packaging-20.9 pluggy-0.13.1 py-1.10.0 pyparsing-2.4.7 pytest-6.2.2 toml-0.10.2
WARNING: You are using pip version 20.2.3; however, version 21.0.1 is available.
You should consider upgrading via the '/home/circleci/.pyenv/versions/3.9.1/bin/python3.9 -m pip install --upgrade pip' command.
Removing intermediate container 13ec3c5acf52
 ---> b8deb04335e7
Successfully built b8deb04335e7
Successfully tagged test2:latest
```
```
testdk|⇒ docker run -it --rm test2:latest pytest --version                           
pytest 6.2.2
```
^ That would still fail of course with your `python -m pip install`. Some quick testing on my own system that is running pyenv results in `python -m pip install` not placing and scripts/execs in the path (maybe my own system isn't setup correctly?). I expect that's because the magic in the `pip` script in the shim is what puts the binary/exec/script in the path of the shim, not in the actual pip package that would be executed directly (skipping the `pip` script in the shim). 

```
circleci@9a3d25552cad:~/project$ python -m pip install pytest
Requirement already satisfied: pytest in /home/circleci/.pyenv/versions/3.9.1/lib/python3.9/site-packages (6.2.2)
Requirement already satisfied: iniconfig in /home/circleci/.pyenv/versions/3.9.1/lib/python3.9/site-packages (from pytest) (1.1.1)
Requirement already satisfied: toml in /home/circleci/.pyenv/versions/3.9.1/lib/python3.9/site-packages (from pytest) (0.10.2)
Requirement already satisfied: py>=1.8.2 in /home/circleci/.pyenv/versions/3.9.1/lib/python3.9/site-packages (from pytest) (1.10.0)
Requirement already satisfied: packaging in /home/circleci/.pyenv/versions/3.9.1/lib/python3.9/site-packages (from pytest) (20.9)
Requirement already satisfied: pluggy<1.0.0a1,>=0.12 in /home/circleci/.pyenv/versions/3.9.1/lib/python3.9/site-packages (from pytest) (0.13.1)
Requirement already satisfied: attrs>=19.2.0 in /home/circleci/.pyenv/versions/3.9.1/lib/python3.9/site-packages (from pytest) (20.3.0)
Requirement already satisfied: pyparsing>=2.0.2 in /home/circleci/.pyenv/versions/3.9.1/lib/python3.9/site-packages (from packaging->pytest) (2.4.7)
WARNING: You are using pip version 20.2.3; however, version 21.0.1 is available.
You should consider upgrading via the '/home/circleci/.pyenv/versions/3.9.1/bin/python -m pip install --upgrade pip' command.
circleci@9a3d25552cad:~/project$ python -m pytest
============================================================ test session starts ============================================================
platform linux -- Python 3.9.1, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
rootdir: /home/circleci/project
collected 0 items                                                                                                                           

=========================================================== no tests ran in 0.00s ===========================================================
circleci@9a3d25552cad:~/project$ pytest
bash: pytest: command not found
```

**Comment by ls-samer-abuthaher** [comment]
Hey! Just to add something here: the reason all shim links broke is because `.bashrc` isn't loaded when you're running install.sh, or any circleci pipeline for that matter, since the `shims` folder was explicitly removed from the path in this PR, depending on `pyenv init` to restore it.

To confirm, try this instead in the docker file: 
```
RUN bash -ic /opt/install.sh
```

Just bringing back shims into the path would be patchy however, so I don't know what's the best way to go around it.

**Comment by dsayling** [comment]
Ok, you're correct, that is the problem with my example test - however, it would still be broken for those that are using this docker image downstream.

```
Sending build context to Docker daemon  3.072kB
Step 1/2 : FROM cimg-python-pr76:latest
 ---> c87f0ea132d0
Step 2/2 : RUN pip install pytest
 ---> Running in 894c332e16dc
+ pip install pytest
/bin/bash: pip: command not found
```

Using the `cimg/python:3.9` as the base would not result in this failure. While we are using pyenv to load the python environment, we don't want users to have to think about the pyenv setup if they don't have to. Sourcing any profile should not be required to setup pyenv correctly. Now, that being said, the `python -m pip install` behavior is something you have to think about, but I don't see any way to relieve that with this PR - the `pyenv rehash` is still needed to get the bin into the path - and I don't see any way around this based on the way `pyenv` behaves.

```
$ docker run --rm -it cimg-python-pr76:latest /bin/bash
circleci@50c93afdff62:~/project$ python -m pip install pytest
Collecting pytest
  Downloading pytest-6.2.2-py3-none-any.whl (280 kB)
     |████████████████████████████████| 280 kB 2.0 MB/s 
Collecting pluggy<1.0.0a1,>=0.12
  Downloading pluggy-0.13.1-py2.py3-none-any.whl (18 kB)
Collecting iniconfig
  Downloading iniconfig-1.1.1-py2.py3-none-any.whl (5.0 kB)
Collecting packaging
  Downloading packaging-20.9-py2.py3-none-any.whl (40 kB)
     |████████████████████████████████| 40 kB 3.3 MB/s 
Collecting toml
  Downloading toml-0.10.2-py2.py3-none-any.whl (16 kB)
Collecting attrs>=19.2.0
  Downloading attrs-20.3.0-py2.py3-none-any.whl (49 kB)
     |████████████████████████████████| 49 kB 3.0 MB/s 
Collecting py>=1.8.2
  Downloading py-1.10.0-py2.py3-none-any.whl (97 kB)
     |████████████████████████████████| 97 kB 3.3 MB/s 
Collecting pyparsing>=2.0.2
  Downloading pyparsing-2.4.7-py2.py3-none-any.whl (67 kB)
     |████████████████████████████████| 67 kB 2.9 MB/s 
Installing collected packages: pluggy, iniconfig, pyparsing, packaging, toml, attrs, py, pytest
Successfully installed attrs-20.3.0 iniconfig-1.1.1 packaging-20.9 pluggy-0.13.1 py-1.10.0 pyparsing-2.4.7 pytest-6.2.2 toml-0.10.2
WARNING: You are using pip version 20.2.3; however, version 21.0.1 is available.
You should consider upgrading via the '/home/circleci/.pyenv/versions/3.9.1/bin/python -m pip install --upgrade pip' command.
circleci@50c93afdff62:~/project$ pytest --version
bash: pytest: command not found
circleci@50c93afdff62:~/project$ pyenv rehash
circleci@50c93afdff62:~/project$ pytest --version
pytest 6.2.2
```
